core/src/commonMain/kotlin/com/xebia/functional/xef/llm/assistants/CachedTool.kt

```diff
@@ -0,0 +1,33 @@
+package com.xebia.functional.xef.llm.assistants
+
+import arrow.fx.coroutines.Atomic
+import arrow.fx.coroutines.timeInMillis
+import kotlin.time.Duration
+import kotlin.time.Duration.Companion.days
+
+abstract class CachedTool<Input, Output>(
+  private val cache: Atomic<MutableList<CachedToolInfo<Input, Output>>>,
```

Comment:
Why not a `MutableMap<Input, CachedToolInfo<Input, Output>>`? That way, the lookups by `Input` would be faster and would solve the issue of allowing duplicates, which currently the list does.